iree/turbine/kernel/wave/index_sequence_analysis.py

```diff
@@ -126,10 +120,135 @@ def has_strided_access(node: fx.Node) -> bool:
                     )
                     for j, dim in enumerate(custom.register_type.symbolic_shape)
                 }
+                ops_to_combine.append(write)
 
+        # Useful to handle write/read dependency
+        custom.replace_all_uses_with(ops_to_combine)
         custom.graph.erase_node(operator)
 
 
+def partition_ops_with_gpr_offsets(trace: CapturedTrace, constraints: list[Constraint]):
+    """
+    This function analyzes the index sequence of reads and writes in a graph.
+    If the reads or writes have incontiguous offsets based on GPR_NUM, we'd
+    need to split these reads/writes appropriately.
+
+    e.g a vector<16xf16> may be owned by lane 0, and lane 16 in this layout:
+    [0, 0, 0, 0, 16, 16, 16, 16, 0, 0, 0, 0, 16, 16, 16, 16].
+
+    With our current glossary, this means we have 2 VGPR "chunks".
+    [0:4) and [8:12) for lane0, and [4:8) and [12:16) for lane16.
+    To the lane it should just look like vector<8xf16>.
+    Hence for this example, we'd need two reads of vector<4xf16> and a couple
+    insert_slices to combine them together into a single vector<8xf16>.
+
+    """
+
+    def has_gpr_offsets(node: fx.Node) -> bool:
+        """
+        Checks for writes on 2d tensors with strided access on a single dimension that
+        read more than a single element.
+        """
+        custom = get_custom(node)
+        if not isinstance(custom, (Read, Write)):
+            return False
+        num_dims_with_gpr = sum(
+            1 for v in custom.index.values() if v.start.has(GPR_NUM)
+        )
+        if num_dims_with_gpr == 1:
+            return True
+        elif num_dims_with_gpr == 0:
+            return False
+        raise NotImplementedError("Currently only handles 1 dim with GPR offset.")
+
+    strided_operators = trace.walk(has_gpr_offsets)
+    for operator in strided_operators:
+        custom = get_custom(operator)
+        simplified_index = {
+            dim: simplify_index(custom.index.get(dim, custom.index[dim]))
+            for dim in custom.index
+        }
+        elements_per_thread = subs_idxc(custom.elements_per_thread)
+        gpr_offsets = [
+            v.start for v in simplified_index.values() if v.start.has(GPR_NUM)
+        ]
+        assert len(gpr_offsets) == 1, "Expected only 1-Dim has gpr offsets"
+        gpr_offset_expr = gpr_offsets[0]
+        gpr_offsets = [
+            gpr_offset_expr.subs({GPR_NUM: i}) for i in range(elements_per_thread)
+        ]
+        gpr_steps = [x - y for x, y in zip(gpr_offsets[1:], gpr_offsets[:-1])]
+
+        # Compute size of chunks by detecting Reads/Writes that has non-contiguous GPR Read/Writes.
```

Comment:
Not sure if this will help, but say you computed

```
gpr_relative_offsets = [x -gpr_offsets[0] for x in gpr_offsets]
```
which gives you
```
 [0, 1, 2, 3, 8, 9, 10, 11, 16, 17, 18, 19, 24, 25, 26, 27]
```
you can group these into
```
[[0, 1, 2, 3], [8, 9, 10, 11], [16, 17, 18, 19], [24, 25, 26, 27]]
```
with this one-liner
```
from itertools import groupby
from operator import itemgetter
[list(map(itemgetter(1), g)) for _, g in groupby(enumerate(gpr_relative_offsets), lambda x: x[0] - x[1])]
```